charts/generic-service/templates/controller.yaml

```
@@ -51,59 +58,56 @@ spec:
     {{- end }}
     type: OnDelete
   {{- else if eq .Values.rollout.strategy "Canary" }}
-    {{- if not .Values.ingress.enabled }}
-      {{ fail "ingress.enabled must be true if rollout.strategy is Canary" }}
-    {{- end }}
-    {{- if not .Values.ingress.domains }}
-      {{ fail "ingress.domains must not be empty if rollout.strategy is Canary" }}
-    {{- end }}
     {{- if eq .Values.rollout.controller "ArgoRollout" }}
     canary:
-      stableService: {{ include "generic-service.fullname" . }}
-      stableMetadata:
-        labels:
-          role: stable
-      canaryService: {{ include "generic-service.fullname" . }}-preview
-      canaryMetadata:
-        labels:
-          role: preview
-      analysis: {{- .Values.rollout.analysis | required "rollout.analysis is required if rollout.strategy is Canary" | toYaml |  nindent 8 }}
-      trafficRouting:
-        {{- if .Values.ingress.istio.enabled }}
-        istio:
-          virtualService:
-            name: {{ include "generic-service.fullname" . }}
-        {{- else if contains "nginx" .Values.ingress.class }}
-        nginx:
-          stableIngress: {{ include "generic-service.fullname" . }}
-        {{- else }}
-          {{ fail "Ingress must use Istio or nginx if rollout.strategy is Canary" }}
+      {{- with .Values.rollout.analysis -}}
+        analysis: {{- . | toYaml | nindent 8 }}
+      {{- end }}
+      {{- if .Values.rollout.useTrafficRouting -}}
+        {{- if not eq .Values.rollout.strategy "Canary" }}
+          {{ fail "Traffic routing is only available if rollout.strategy is Canary" }}
+        {{- end }}
+        stableService: {{ include "generic-service.fullname" . }}
+        stableMetadata:
+          labels:
+            role: stable
+        canaryService: {{ include "generic-service.fullname" . }}-preview
+        canaryMetadata:
+          labels:
+            role: preview
+        trafficRouting:
+          {{- if .Values.ingress.istio.enabled }}
+            istio:
+              virtualService:
+                name: {{ include "generic-service.fullname" . }}
+          {{- else if eq .Values.ingress.class "nginx" }}
+            nginx:
+              stableIngress: {{ include "generic-service.fullname" . }}
+          {{- else }}
+            {{ fail "Ingress must use Istio or nginx if rollout.strategy is Canary" }}
+          {{- end }}
+        {{- end }}
+        {{- with .Values.rollout.steps }}
+        steps: {{- . | toYaml | nindent 8 }}
```

Indent seems to be off:

```suggestion
        steps: {{- . | toYaml | nindent 10 }}
```